File: 0307 - Range Sum Query - Mutable/solution.py
class NumArray:

    def __init__(self, nums: List[int]):
        self.sz = len(nums)
        self.A = [0]*(self.sz+1)
        for i, v in enumerate(nums):
            self.update(i, v)
            
    def query(self, idx: int) -> int:
        idx += 1; res = 0
        while idx > 0:
            res += self.A[idx]
            idx -= idx&-idx
        return res

    def update(self, idx: int, val: int) -> None:
        dval = val-(self.query(idx)-self.query(idx-1))
        idx += 1
        while idx <= self.sz:
            self.A[idx] += dval
            idx += idx&-idx

    def sumRange(self, left: int, right: int) -> int:
        return self.query(right)-self.query(left-1)


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)

File: 0307 - Range Sum Query - Mutable/solution.c
typedef struct {
    int sz, *A;
} NumArray;

NumArray* numArrayCreate(int* nums, int numsSz) {
    NumArray *res = (NumArray*)malloc(sizeof(NumArray));
    res->sz = numsSz;
    res->A = (int*)calloc(res->sz+1, sizeof(int));
    for(int i = 0; i < res->sz; i++)
        numArrayUpdate(res, i, nums[i]);
    return res;
}

int numArrayQuery(NumArray* obj, int idx){
    int res = 0;
    for(int i = idx+1; i > 0; i -= i&-i)
        res += obj->A[i];
    return res;
}

void numArrayUpdate(NumArray* obj, int idx, int val) {
    int dval = val-(numArrayQuery(obj, idx)-numArrayQuery(obj, idx-1));
    for(int i = idx+1; i <= obj->sz; i += i&-i)
        obj->A[i] += dval;
}

int numArraySumRange(NumArray* obj, int left, int right) {
    return numArrayQuery(obj, right)-numArrayQuery(obj, left-1);
}

void numArrayFree(NumArray* obj) {
    free(obj->A);
}

/**
 * Your NumArray struct will be instantiated and called as such:
 * NumArray* obj = numArrayCreate(nums, numsSize);
 * numArrayUpdate(obj, index, val);
 
 * int param_2 = numArraySumRange(obj, left, right);
 
 * numArrayFree(obj);
*/